File: Software/NodeJS/src/components/servo.js
// https://www.dexterindustries.com/GoPiGo/
// https://github.com/DexterInd/GoPiGo3
//
// Copyright (c) 2017 Dexter Industries
// Released under the MIT license (http://choosealicense.com/licenses/mit/).
// For more information see https://github.com/DexterInd/GoPiGo3/blob/master/LICENSE.md
//

const ValueError = require('../errors/valueError');

const Sensor = require('./sensor');

/**
 * Wrapper to control the Servo Motors on the GPG3.
    Allows you to rotate the servo by feeding in the angle of rotation.
    Connect the Servo to the Servo1 and Servo2 ports of GPG3.
 */
class Servo extends Sensor {
    PULSE_WIDTH_RANGE = 1850;
    // Pulse width range in us corresponding to 0 to 180 degrees

    constructor(port = 'SERVO1', gpg) {
        try {
            super(port, 'OUTPUT', gpg);
            this.setDescriptor('GoPiGo3 Servo');
        } catch (err) {
            throw new ValueError('GoPiGo3 Servo not found');
        }
    }

    /**
     * This calculation will vary with servo and is an approximate angular movement of the servo
        Pulse Width varies between 575us to 24250us for a 60KHz Servo Motor which
        rotates between 0 to 180 degrees.

        0 degree ~= 575us
        180 degree ~= 2425us
        Pulse width Range= 2425-575 =1850
        => 1 degree rotation requires ~= 10.27us
     * @param {*} servoPosition
     */
    rotateServo(servoPosition) {
        servoPosition = servoPosition > 180 ? 180 : servoPosition;
        servoPosition = servoPosition < 0 ? 0 : servoPosition;

        const pulseWidth = Math.round(
            (1500 - (this.PULSE_WIDTH_RANGE / 2)) +
            ((this.PULSE_WIDTH_RANGE / 180) * servoPosition)
        );
        this.gpg.setServo(this.getPortId(), parseInt(pulseWidth, 0));
    }

    /**
     *
     */
    resetServo() {
        this.gpg.setServo(this.getPortId(), 0);
    }

}

module.exports = Servo;

File: Software/Java_new/main/java/com/dexterindustries/gopigo3driver/Servo.java
package com.dexterindustries.gopigo3driver;

import java.io.IOException;

public class Servo extends Sensor {
	private static int PULSE_WIDTH_RANGE = 1850;
	
	/**
	 * Constructor for initializing a servo object
	 * @param port can be either "SERVO1" or "SERVO2"
	 * @param gpg an instance of a gopigo3 object
	 * @param use_mutex when using multiple sensors in parallel, mutexes should be enabled.
	 * @throws IOException
	 * @throws InterruptedException
	 */
	public Servo(String port, GoPiGo3 gpg, boolean use_mutex) throws IOException, InterruptedException {
		super(port, "OUTPUT", gpg, use_mutex);
		set_descriptor("SERVO");
	}
	
	/**
	 * Rotates the servo to a specific angle.
	 * 
	 * The pulse width here is in microseconds and varies int he following way:
	 * 0 degrees = 575 uS.
	 * 180 degrees = 24250 uS.
	 * Thankfully, this function accepts inputs in degrees.
	 * @param servo_position the target position in degrees.
	 * @throws IOException
	 */
	public void rotate_servo(double servo_position) throws IOException {
		if(servo_position > 180) {
			servo_position = 180;
		}
		if(servo_position < 0) {
			servo_position = 0;
		}
		
		int pulsewidth = (int)((1500 - (PULSE_WIDTH_RANGE/2)) + ((PULSE_WIDTH_RANGE/180) * servo_position));
		gpg.set_servo(get_port_id(), pulsewidth);
	}
	
	/**
	 * Resets the servo straight ahead.
	 * @throws IOException
	 */
	public void reset_servo() throws IOException {
		rotate_servo(90);
	}
	
	/**
	 * Disables the servo. The purpose of this command is so that if you try to turn the servo manually, it won't resist you.
	 * @throws IOException
	 */
	public void disable_servo() throws IOException {
		gpg.set_servo(get_port_id(), 0);
	}
}
